### DISCUSSION TITLE
FragmentToSmiles - How to include all bonds of the included Atoms?

### DISCUSSION THREAD

**Discussion by JanoschMenke** [rdkit/rdkit#6197] [Q&A] [answered]
Hey Guys,

I am trying to extract the SMARTS for molecules based on atom indices provided by the user. I can extract the SMARTS like so:

```python
from rdkit.Chem import AllChem as Chem
rdkit.Chem import Draw 
from rdkit.Chem.Draw import IPythonConsole
IPythonConsole.drawOptions.addAtomIndices = True 
IPythonConsole.drawOptions.addBondIndices = True
mol = Chem.MolFromSmiles("CC(=O)NCCc1c[nH]c2ccc(OC)cc12")
mol

Chem.MolFragmentToCXSmarts(mol,[1,2,3,4,5,6],bondsToUse=[0,1,2,3,4,5,6,16])
```

which we result in the SMARTS:
`'[#6](=[#8])-[#7]-[#6]-[#6]-[#6]'`

This is not completely correct as it ignores the additional Bonds that I specified in the `bondsToUse` parameter (index 0,6,16). Which are the bonds of atom 1 and 6 that connect the fragment to atoms not included in the fragment.
Ideally I would like to see a SMARTS like so:

`'[*]-[#6](=[#8])-[#7]-[#6]-[#6]-[#6](-[*])-[*]'`

which includes all bond information of the included atoms.

Is there a way to be able achieve this?

Greetings and many thanks in advance.

**Comment by greglandrum** [comment]
Hi @JanoschMenke. The various `MolFragmentToX()` functions only output the atoms and bonds that you as them to output.
The code for drawing Morgan fingerprint environments shows to to extract environments + the atoms around them as '*' atoms into a molecule. That code then draws the molecule, but you could just as easily generate a SMILES or SMARTS from it:
https://github.com/rdkit/rdkit/blob/6cc277f9611d31d27de092475d9cd44d13112de8/rdkit/Chem/Draw/__init__.py#L735

**Reply by JanoschMenke** [reply]
In case someone else is in dire need of such function. I did not perform any tests to check how well this function will actually work.

```python
  from rdkit.Chem import AllChem as Chem
  from rdkit.Chem import Draw 
  from rdkit.Chem.Draw import IPythonConsole
  IPythonConsole.drawOptions.addAtomIndices = True 
  IPythonConsole.drawOptions.addBondIndices = True
  
  mol = Chem.MolFromSmiles("CC(=O)NCCc1c[nH]c2ccc(OC)cc12")
  mol
  
  Chem.MolFragmentToCXSmarts(mol,[1,2,3,4,5,6],bondsToUse=[0,1,2,3,4,5,6,16])
   
  def MolFragmentToSmartsEnlarged(mol, atomids):
      atomids = set(atomids)
      enlargedEnv = set()
      for atom in atomids:
          a = mol.GetAtomWithIdx(atom)
          for b in a.GetBonds():
              enlargedEnv.add(b.GetIdx())
      enlargedEnv = list(enlargedEnv)
      amap = {}
      submol=Chem.PathToSubmol(mol, enlargedEnv, atomMap=amap)
      for i in (set([key for key in amap])-atomids):
          submol.GetAtomWithIdx(amap[i]).SetAtomicNum(0)
      return Chem.MolToSmarts(submol).replace("#0", "*")    
  
  MolFragmentToSmartsEnlarged(mol, [1,2,3,4,5,6] )
```

**Comment by JanoschMenke** [comment]
Hi Greg,

awesome thanks. I will try that.

**Comment by JanoschMenke** [comment, marked as answer]
```python
def MolFragmentToSmartsEnlarged(mol, atomids):
      atomids = set(atomids)
      enlargedEnv = set()
      for atom in atomids:
          a = mol.GetAtomWithIdx(atom)
          for b in a.GetBonds():
              enlargedEnv.add(b.GetIdx())
      enlargedEnv = list(enlargedEnv)
      amap = {}
      submol=Chem.PathToSubmol(mol, enlargedEnv, atomMap=amap)
      for i in (set([key for key in amap])-atomids):
          submol.GetAtomWithIdx(amap[i]).SetAtomicNum(0)
      return Chem.MolToSmarts(submol).replace("#0", "*")  
```
